### ISSUE TITLE
Special treatment needed when training first 30 moves of a game

### ISSUE THREAD

**Issue by billyswong** [gcp/leela-zero#250]
The training data is fed in the format of probability maps for game states. We encourage probability maps that lead to a game win and discourage probability maps that lead to a game lose. However, since the first 30 moves are played probabilistically after computing the map, the game move is often different from the strongest move the map suggest. Any training based on those data points will be like training on noise since the game win/lose is not exactly caused by the map suggestion.

We need to either:
1) discard data points from training whenever a data point doesn't match the game move; or
2) train on game moves instead for those cases

Correct me if it has already been implemented.

**Comment by isty2e** [comment]
AFAIK the NN is trained with the probability map obtained with MCTS for each board state, so as long as MCTS is working fine I don't think we need to do that.

**Comment by billyswong** [comment]
From AGZ paper:
> For the first 30 moves of each game, the temperature is set to τ = 1; this selects moves proportionally to their visit count in MCTS, and ensures a diverse set of positions are encountered. Additional exploration is achieved by adding Dirichlet noise to the prior probabilities in the root node s_0

For the later moves, MCTS map result reflect the game move exactly but not so for the first 30 moves. Thus Leela Zero's "NN is trained with the probability map obtained with MCTS for each board state" is a problem.

**Comment by isty2e** [comment]
Yes, I know that the first 30 moves are noised. The purpose of the randomization is to have diverse board positions (training set input). Though the moves played are not optimal, it is okay because they are not the output of the NN, but rather the inputs. As long as MCTS itself is okay, the MCTS probability map (training set output/target) is also fine, so the quality of training is unaffected.

In short, by randomization we only get more diverse inputs without deteriorating the training set output.

**Comment by billyswong** [comment]
So it's okay even though the map mismatch the game move + the game result, making it somewhat a "incorrect" data point?

**Comment by isty2e** [comment]
Theoretically it can malfunction a bit for the value net, but has no effect on the policy net, I think.

**Comment by earthengine** [comment]
One of the idea in my mind is to change rule as any pass shall be equal to a resign unless the board is 10% filled.

In practice, this change on the rule should have no impact to the result, but it will remove the endless "pass" in self plays.

**Comment by billyswong** [comment]
@earthengine It's not about early pass. It's about early pass or whatever bad moves in early game can't be trained away effectively in current scheme.

Because the probability map does not necessarily match the game move, it is fairly likely that the map suggests a losing move and the game is won because the AI chose another move probabilistically; or in reverse the map suggests a winning move and the game is lost because the AI chose otherwise again. In the current training algorithm, Leela Zero unilaterally assumes it is the map that cause the game result while it is not. One can foresee that the training on how to handle the first 30 moves will be like random walk; any improvement has to be stumbled on by luck instead of actually learned by proper data.

This could be a reason why Leela Zero still play a pass as white, could be not. But my point is we are currently hindering the training effectiveness unnecessarily.

Your "pass equal resign in early game" idea doesn't work because this will inject new human knowledge into the experiment.

**Comment by gcp** [comment]
The probability map is for the policy (head of the... ) network, not the value network. The policy network doesn't care who won or not.

**Comment by billyswong** [comment]
So the value head train on game moves only? Thanks for clarification.
